packages/grafana-data/src/dataframe/DataFrameJSON.ts

```diff
@@ -0,0 +1,153 @@
+   */
+  schema?: DataFrameSchema;
+
+  /**
+   * The data is a columnar store that matches fields defined by schema.
+   */
+  data?: any[][];
+
+  /**
+   * Since JSON can not encode some important numeric constants, these
+   * replacements are applied after reading JSON over-the-wire.  This addition allows
+   * us to represent values for NaN, Inf, -Inf, and undefined
+   */
+  replaced?: FieldValueReplacements[];
+}
+
+/**
+ * The JSON transfer object for DataFrames.  Values are stored in simple JSON
+ *
+ * @alpha
+ */
+export interface DataFrameSchema {
+  /**
+   * Matches the query target refId
+   */
+  refId?: string;
+
+  /**
+   * Initial response global metadata
+   */
+  meta?: QueryResultMeta;
+
+  /**
+   * Frame name
+   */
+  name?: string;
+
+  /**
+   * Field definition without any metadata
+   */
+  fields?: FieldSchema[];
+}
+
+/**
+ * Field object passed over JSON
+ *
+ * @alpha
+ */
+export interface FieldSchema {
+  name: string; // The column name
+  type?: FieldType;
+  config?: FieldConfig;
+  labels?: Labels;
+}
+
+/**
+ * Since JSON does not support encoding some numeric types, we need to replace them on load.
+ * While some system solve this by encoding everythign as strings, this approach lets is
+ * efficiently communicate which values needed special encoding.
+ *
+ * @alpha
+ */
+export interface FieldValueReplacements {
+  NaN?: number[];
+  Undef?: number[]; // Missing because of absense or join
+  Inf?: number[];
+  NegInf?: number[];
+}
+
+const replacements: Record<keyof FieldValueReplacements, any> = {
+  Inf: Infinity,
+  NegInf: -Infinity,
+  Undef: undefined,
+  NaN: NaN,
+};
+
+export function applyFieldValueReplacements(replaced: FieldValueReplacements, col: any[]) {
+  if (!replaced || !col) {
+    return;
+  }
+  for (const key in replaced) {
+    const val = replacements[key as keyof FieldValueReplacements];
+    for (const idx of replaced[key as keyof FieldValueReplacements]!) {
+      col[idx] = val;
+    }
+  }
+}
+
+function guessFieldType(name: string, values: any[]): FieldType {
+  for (const v of values) {
+    if (v != null) {
+      return guessFieldTypeFromNameAndValue(name, v);
+    }
+  }
+  return FieldType.other;
+}
+
+/**
+ * NOTE: data in the original array will be replaced with values from the DataFrameReplacements property
+ *
+ * @alpha
+ */
+export function dataFrameFromJSON(dto: DataFrameJSON): DataFrame {
+  const { schema, replaced, data } = dto;
+  if (!schema || !schema.fields) {
+    throw new Error('JSON needs a fields definition');
+  }
+
+  // Find the longest field length
+  let length = 0;
+  if (data) {
+    for (const f of data) {
+      let flen = f.length;
+      if (flen && flen > length) {
+        length = flen;
+      }
+    }
+  }
+
+  const fields = schema.fields.map((f, index) => {
+    let r: FieldValueReplacements = {};
+    let buffer = data ? data[index] : [];
+    buffer.length = length; // will pad with undefined
```

Comment:
if this is an array, i think i't best to actually fill it with `undefined`, otherwise you get a sparse often de-optimized array by doing this.

compare the console output of

```js
let x = [];
x.length = 100;
console.log(x);

let y = [];
y.length = 100;
y.fill(undefined, 0, 100);
console.log(y);
```

Reply:
this started out without the fill, and the perf difference was quite large:

https://github.com/leeoniya/uPlot/blob/6ad1cb86e85672604515045abca3c891aab94086/src/utils.js#L377

Reply:
interesting -- yes, the intention here is that they end up the same length in whatever the most efficient structure is.